#! /bin/sh -e

# gdc-driver-zlib.dpatch by Arthur Loiret <arthur.loiret@gmail.com>
# DP: Update the gdc driver to use the up-to-date system zlib.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        #cd ${dir}gcc && autoconf
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0


diff -ruN gcc/d/d-spec.c gcc/d/d-spec.c
--- gcc/d/d-spec.c	2007-09-27 17:29:42.000000000 +0200
+++ gcc/d/d-spec.c	2007-09-27 17:34:51.000000000 +0200
@@ -87,7 +87,7 @@
     /* The number of arguments being added to what's in argv, other than
        libraries.  We use this to track the number of times we've inserted
        -xc++/-xnone.  */
-    int added = 0;
+    int added = 1; /* -lz */
 
     /* Used to track options that take arguments, so we don't go wrapping
        those with -xc++/-xnone.  */
@@ -366,6 +366,10 @@
 	    arglist[j++] = saw_profile_flag ? LIBPHOBOS_PROFILE : LIBPHOBOS;
 	    added_libraries++;
 	    need_pthreads = 1;
+
+	    /* Use the up-to-date system zlib with libphobos */
+	    arglist[j++] = "-lz";
+	    added_libraries++;
 	}
     if (saw_math)
 	arglist[j++] = saw_math;
diff -ruN gcc/d/phobos/etc/c/zlib.d gcc/d/phobos/etc/c/zlib.d
--- gcc/d/phobos/etc/c/zlib.d	2007-09-27 17:29:42.000000000 +0200
+++ gcc/d/phobos/etc/c/zlib.d	2007-09-27 17:38:16.000000000 +0200
@@ -41,7 +41,7 @@
 
 extern (C):
 
-char[] ZLIB_VERSION = "1.2.1";
+char[] ZLIB_VERSION = "1.2.3.3";
 
 /*
      The 'zlib' compression library provides in-memory compression and
diff -ruN gcc/d/phobos2/etc/c/zlib.d gcc/d/phobos2/etc/c/zlib.d
--- gcc/d/phobos2/etc/c/zlib.d	2007-09-27 17:29:44.000000000 +0200
+++ gcc/d/phobos2/etc/c/zlib.d	2007-09-27 17:38:26.000000000 +0200
@@ -41,7 +41,7 @@
 
 extern (C):
 
-const char[] ZLIB_VERSION = "1.2.1";
+const char[] ZLIB_VERSION = "1.2.3.3";
 
 /*
      The 'zlib' compression library provides in-memory compression and
--- gcc/d/phobos/Makefile.am	2007-11-10 19:44:58.000000000 +0100
+++ gcc/d/phobos/Makefile.am	2007-11-10 19:45:46.000000000 +0100
@@ -177,7 +177,7 @@
 std/stream.t.o: std/stream.d $(D_PREREQ_SRCS)
 	$(GDC) -o $@ $(ALL_DFLAGS) -fdeprecated -I $(srcdir) -I $(srcdir)/internal/gc -I ./$(host_alias) -c $<
 
-ALL_PHOBOS_OBJS = $(D_EXTRA_OBJS) $(GCC_OBJS) $(MAIN_OBJS) $(ZLIB_OBJS) $(GC_OBJS) $(WEAK_OBJS)
+ALL_PHOBOS_OBJS = $(D_EXTRA_OBJS) $(GCC_OBJS) $(MAIN_OBJS) $(GC_OBJS) $(WEAK_OBJS)
 
 libgphobos.a : $(ALL_PHOBOS_OBJS) $(CMAIN_OBJS)
 	$(AR) -r $@ $(ALL_PHOBOS_OBJS) $(CMAIN_OBJS)
--- gcc/d/phobos2/Makefile.am	2007-11-10 19:45:13.000000000 +0100
+++ gcc/d/phobos2/Makefile.am	2007-11-10 19:46:02.000000000 +0100
@@ -177,7 +177,7 @@
 std/stream.t.o: std/stream.d $(D_PREREQ_SRCS)
 	$(GDC) -o $@ $(ALL_DFLAGS) -fdeprecated -I $(srcdir) -I $(srcdir)/internal/gc -I ./$(host_alias) -c $<
 
-ALL_PHOBOS_OBJS = $(D_EXTRA_OBJS) $(GCC_OBJS) $(MAIN_OBJS) $(ZLIB_OBJS) $(GC_OBJS) $(WEAK_OBJS)
+ALL_PHOBOS_OBJS = $(D_EXTRA_OBJS) $(GCC_OBJS) $(MAIN_OBJS) $(GC_OBJS) $(WEAK_OBJS)
 
 libgphobos2.a : $(ALL_PHOBOS_OBJS) $(CMAIN_OBJS)
 	$(AR) -r $@ $(ALL_PHOBOS_OBJS) $(CMAIN_OBJS)
